pylabrobot/config/service/file.py

```diff
@@ -0,0 +1,48 @@
+from pathlib import Path
+from typing import IO, Union, Dict
+
+from pylabrobot.config.config import Config
+from pylabrobot.config.service.reader import ConfigReader
+from pylabrobot.config.service.writer import ConfigWriter
+
+
+class FileReader(ConfigReader[str]):
+  """A ConfigReader that reads from a file"""
+
+  def __init__(self, format_reader: ConfigReader[IO]):
+    self.format_reader = format_reader
+
+  def read(self, r: str) -> Config:
+    """Read a Config object from a file."""
+    with open(r, self.format_reader.open_mode) as f:
+      return self.format_reader.read(f)
+
+
+class FileWriter(ConfigWriter[str]):
+  """A ConfigWriter that writes to a file"""
+
+  def __init__(self, format_writer: ConfigWriter[IO]):
+    self.format_writer = format_writer
+
+  def write(self, w: str, cfg: Config):
+    """Write a Config object to a file."""
+    with open(w, self.format_writer.open_mode) as f:
+      return self.format_writer.write(f, cfg)
+
+
+class MultiReader(ConfigReader[str]):
+  """A ConfigReader that can read from multiple formats"""
+
+  def __init__(self, reader_map: Dict[str, ConfigReader]):
+    self.reader_map = reader_map
+
+  def read(self, r: Union[str, Path]) -> Config:
+    """Read a Config object from a file."""
+    if isinstance(r, str):
+      r = Path(r)
+    file_ext = r.suffix.lstrip(".")
+    reader = self.reader_map.get(file_ext)
+    if reader is None:
+      raise ValueError(f"Unknown file extension: {file_ext}")
+    with open(r, reader.open_mode) as f:
+      return reader.read(f)
```

Comment:
should this exist in service.file.py?

Reply:
I named this poorly -- it is intended for reading from files which could be multiple formats. Right now it is just used with the .ini reader, but can quickly be updated to support config with .json or .yaml.

I separated the file opening, then reading so that i can support from any text stream, i.e. for configuration via http instead of from a file. Could definitely simplify if preferred. I write a lot of Go, so am used to using io.Reader/io.Writer.

Reply:
Then my question would be why `FileReader` is needed? It seems that the `IniReader` is able to read files by itself. The only exception is `test_file_reader_writer`, where the opened file is passed.

It would make sense to either:
1. allow ConfigReaders to read files, which would complicate reading from http/other sources
2. always have ConfigReaders read from an io stream, but then FileReader will not be a ConfigReader I think

2 is more powerful, but also more code.